[now-node] Use `node-file-trace` approach as `ncc` alternative

guybedford opened the pull request:
This integrates the filtering from https://github.com/zeit/node-file-trace into the builder.

* `inputFiles` are all traced for dependencies, using the same logic (and all the tests) from ncc and the asset relocator loader.
* Any files that contain `import` or `export` syntax are compiled into CommonJS using Babel.
* Any `.ts` files are compiled using TypeScript with full typechecking errors
* Source maps are supported for ES modules and TypeScript with an inbuilt version of the sourcemap-register library as used in ncc

We can likely still tweak the logging output and exact handling of errors in Babel / TypeScript.

**This PR should be fully backwards compatible with the ncc approach.**

- Fixes #348
- Fixes #645

styfle commented on packages/now-node/package.json:
```diff
@@ -12,16 +12,17 @@
     "@types/node": "*",
-    "@zeit/ncc": "0.18.5",
-    "@zeit/ncc-watcher": "1.0.3"
+    "@zeit/ncc": "0.20.4",
```
I think you can remove `@zeit/ncc` right?

guybedford replied:
It's still used to build the helpers. Although perhaps that makes it a `devDependency`? I'm not sure.

styfle replied:
Good point! It should be `devDependencies` instead.

styfle commented on packages/now-node/package.json:
```diff
@@ -12,16 +12,17 @@
   "files": [
-    "dist"
+    "dist",
+    "node-file-trace"
```
I don't think you need to publish `node-file-trace` any longer

guybedford replied:
thanks

styfle commented on .eslintignore:
```diff
@@ -15,3 +15,4 @@
 /packages/now-ruby/dist/*
 /packages/now-static-build/dist/*
 /packages/now-static-build/test/fixtures/**
+/packages/now-node/node-file-trace/**
```
I think you can remove this line now

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -128,45 +76,65 @@ async function compile(
+  const { fileList } = await nodeFileTrace([...inputFiles], {
+    base: workPath,
+    filterBase: true,
+    ignore: config && (<any>config).excludeFiles,
```
```suggestion
    ignore: config && config.excludeFiles,
```

guybedford replied:
Apparently `excludeFiles` isn't in the type definitions!?

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -19,6 +19,7 @@ import {
+import { readFileSync } from 'fs';
 
 interface CompilerConfig {
   includeFiles?: string | string[];
```
```suggestion
  includeFiles?: string | string[];
  excludeFiles?: string | string[];
```

guybedford replied:
Ahh.

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -128,45 +76,65 @@ async function compile(
+      // null represents a not found
+      if (cached === null) return null;
+      try {
+        const source = readFileSync(path);
```
Why is this synchronous? Can you use the async `readFile` instead?

guybedford replied:
No, the fs hook for node-file-trace is explicitly sync for architecture reasons.

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -128,45 +76,65 @@ async function compile(
+      if (cached === null) return null;
+      try {
+        const source = readFileSync(path);
+        // TODO: set file mode here
```
You can read the `stats.mode` from `path`

guybedford replied:
Yeah it's just a statSync line needed I think.

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -245,12 +296,4 @@ export async function build({
-    ...(await glob('package-lock.json', workPath)),
-    ...(await glob('yarn.lock', workPath)),
-  };
-}
```
Why did you remove `prepareCache`?

It is necessary, see https://zeit.co/docs/v2/deployments/builders/developer-guide

guybedford replied:
It says in the docs it is optional. I thought crawling all of node_modules may actually have worse performance than just using the trace we have?

guybedford replied:
I've added this back for now, and we can investigate improvements in due course.

styfle commented:
@guybedford I thought this would fix #348 but I tried this builder using [the sharp repo here](https://github.com/friedjoff/now-sharp-example) and it fails with `Error: /var/task/node_modules/sharp/build/Release/sharp.node: ELF file's phentsize not the expected size`.

styfle commented:
@guybedford This also fails to run the typescript project [og-image](https://github.com/zeit/og-image) with error `Cannot find module './parser'`.

styfle commented on .eslintignore:
```diff
@@ -14,4 +14,4 @@
 /packages/now-ruby/dist/*
 /packages/now-static-build/dist/*
-/packages/now-static-build/test/fixtures/**
+/packages/now-static-build/test/fixtures/**
```
Please add a newline

styfle commented on packages/now-node/build.sh:
```diff
@@ -27,3 +27,8 @@ rm dist/helpers.js
+# build source-map-support/register for source maps
+ncc build node_modules/source-map-support/register -o dist/source-map-support
+mv dist/source-map-support/index.js dist/source-map-support.js
+rm -rf dist/source-map-support
```
Please add a newline

guybedford commented:
@styfle thanks for the tests!

> I thought this would fix #348 but I tried this builder using the sharp repo here and it fails with Error: /var/task/node_modules/sharp/build/Release/sharp.node: ELF file's phentsize not the expected size.

I just fixed this one and have verified it is working now.

> This also fails to run the typescript project og-image with error Cannot find module './parser'.

This was actually due to a valid typing issue in that project here - https://github.com/zeit/og-image/blob/master/src/parser.ts#L46 where `Theme` is not defined.

I've fixed and confirmed that is working, although now I'm getting a Pupeteer bug. Will look into this too.

styfle commented on yarn.lock:
```diff
@@ -1183,18 +1305,33 @@
-    "@zeit/ncc" "^0.18.5"
-
-"@zeit/ncc@0.18.5", "@zeit/ncc@^0.18.5":
+"@zeit/ncc@0.18.5":
```
Please run `yarn` from the root because we should see ncc@0.18.5 disappear

styfle commented on packages/now-node/test/fixtures/12-stack-trace-esm/index.js:
```diff
@@ -0,0 +1,10 @@
+export default function handler(req, res) {
+  try {
+    if (req) {
+      throw new Error('Should throw');
```
```suggestion
      req.notdefined.something;
```

Let's use the same code from #645 to prove it is fixed.

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -127,46 +86,140 @@ async function compile(
     }
   }
 
+  /*console.log(
```
Please remove unused comments

guybedford replied:
I find these very useful for debugging - is there a `log` option we can guard these under?

styfle replied:
You can use `if (config.debug) console.log()`

guybedford replied:
So I can just add a `debug` to the `CompilerConfig` type?

styfle replied:
Correct

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -127,46 +86,140 @@ async function compile(
+  });
+
+  // console.log('traced files:');
+  // console.log('\t' + fileList.join('\n\t'));
```
Please remove unused comments

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -127,46 +86,140 @@ async function compile(
+    // Rename .ts -> .js (except for entry)
+    if (path !== entrypoint && tsCompiled.has(path))
+      preparedFiles[path.slice(0, -3) + '.js'] = entry;
+    else preparedFiles[path] = entry;
```
Can you change the `if` and `else` statements to always wrap with curly brackets? (I should probably update our linter to check for this)

guybedford replied:
Sure.

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -127,46 +86,140 @@ async function compile(
+      const source = readFileSync(resolved);
+      const stats = statSync(resolved);
+      if (stats.isSymbolicLink()) {
+        // TODO: handle asset symlinks
```
Is this TODO still relevant?

guybedford replied:
Yup.

styfle replied:
What is the use case here? We have sharp working properly with symlinks so whats left?

guybedford replied:
Right, so it's not causing any issues yet, but it could lead to some obscure edge cases in theory.
